### PR TITLE
fix: Don't serve directories as static files

### PR DESCRIPTION
**Pull request by TatuLund**
Adopted from (Java 8 -> Java 6 conversion and simplification)

https://github.com/vaadin/framework/pull/12325

### CODE REVIEW

File: server/src/main/java/com/vaadin/server/VaadinServlet.java [hunk 1]
```diff
@@ -1201,6 +1209,88 @@ protected boolean isAllowedVAADINResourceUrl(HttpServletRequest request,
...
+        // If not a jar check if a file path directory.
+        File file = new File(resourceURI);
+        return "file".equals(resource.getProtocol()) && file.isDirectory();
+    }
+
+    private ZipEntry findEntry(ZipInputStream in, String name)
+            throws IOException {
+        ZipEntry entry = null;
+        while ((entry = in.getNextEntry()) != null) {
+            System.out.println("File: " + entry.getName());
```
[the remark this answers is not in the document]
Comment (TatuLund) [reply server/src/main/java/com/vaadin/server/VaadinServlet.java]: Done

File: server/src/main/java/com/vaadin/server/VaadinServlet.java [hunk 2]
```diff
@@ -1201,6 +1209,88 @@ protected boolean isAllowedVAADINResourceUrl(HttpServletRequest request,
...
+                ZipEntry entry = null;
+                if (pathOfWar == null) {
+                    jar = new ZipFile(pathOfJar);
+                    entry = jar.getEntry(pathInJar);
+                } else {
+                    war = new ZipFile(pathOfWar);
+                    ZipEntry jarEntry = war.getEntry(pathOfJar);
+                    InputStream in = war.getInputStream(jarEntry);
+                    ZipInputStream stream = new ZipInputStream(in);
+                    entry = findEntry(stream, pathInJar);
```
[the remark this answers is not in the document]
Comment (TatuLund) [reply server/src/main/java/com/vaadin/server/VaadinServlet.java:1260]: Fixed

File: server/src/main/java/com/vaadin/server/VaadinServlet.java [hunk 3]
```diff
@@ -1201,6 +1209,88 @@ protected boolean isAllowedVAADINResourceUrl(HttpServletRequest request,
...
+                }
+                if (entry != null) {
+                    isDirectory = entry.isDirectory();
+                }
+                if (war != null) {
+                    war.close();
+                }
+                if (jar != null) {
+                    jar.close();
+                }
```
[the remark this answers is not in the document]
Comment (TatuLund) [reply server/src/main/java/com/vaadin/server/VaadinServlet.java]: Fixed

File: server/src/main/java/com/vaadin/server/VaadinServlet.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (caalador) [reply server/src/main/java/com/vaadin/server/VaadinServlet.java:1260]: Would perhaps be cleaner if having them in different methods e.g. `ZipEntry getJarEntry(String ofJar, String inJar)` & `ZipEntry getWarEntry(String ofWar, String ofJar, String inJar)`

File: server/src/main/java/com/vaadin/server/VaadinServlet.java [hunk 4]
```diff
@@ -1201,6 +1209,95 @@ protected boolean isAllowedVAADINResourceUrl(HttpServletRequest request,
...
+                    } catch (IOException e) {
+                    }
+                }
+            }
+        }
+
+        // If not a jar check if a file path directory.
+        File file = new File(resourceURI);
+        return "file".equals(resource.getProtocol()) && file.isDirectory();
+    }
```
Comment (caalador) [inline server/src/main/java/com/vaadin/server/VaadinServlet.java:1288]:
```suggestion

            try {
                ZipEntry entry = null;
                if (pathOfWar == null) {
                    entry = getJarEntry(pathOfJar, pathInJar);
                } else {
                    entry = getWarEntry(pathOfWar, pathOfJar, pathInJar);
                }
                if (entry != null) {
                    isDirectory = entry.isDirectory();
                }
                return isDirectory;
            } catch (IOException e) {
                return false;
            }
        }

        // If not a jar check if a file path directory.
        File file = new File(resourceURI);
        return "file".equals(resource.getProtocol()) && file.isDirectory();
    }

    private ZipEntry getWarEntry(String pathOfWar, String pathOfJar, String pathInJar) throws IOException {
        ZipFile war = null;
        ZipInputStream stream = null;
        try {
            war = new ZipFile(pathOfWar);
            ZipEntry jarEntry = war.getEntry(pathOfJar);
            InputStream in = war.getInputStream(jarEntry);
            stream = new ZipInputStream(in);
            return findEntry(stream, pathInJar);
        } finally {
            if(stream != null) {
                stream.close();
            }
            if(war != null) {
                war.close();
            }
        }
    }

    private ZipEntry getJarEntry(String pathOfJar, String pathInJar) throws IOException {
        ZipFile jar = null;
        try {
            jar = new ZipFile(pathOfJar);
            return jar.getEntry(pathInJar);
        }finally {
            if(jar != null) {
                jar.close();
            }
        }
    }
```

Comment (TatuLund) [reply server/src/main/java/com/vaadin/server/VaadinServlet.java]: Done. I added clarifying comments too.

**Review by caalador — APPROVED** [review]

**Pull request merged**